complex/src/lib.rs

```diff
@@ -740,6 +742,141 @@ impl<T> fmt::Binary for Complex<T> where
     }
 }
 
+impl<T> FromStr for Complex<T> where
+    T: FromStr + Num + PartialOrd + Clone
+{
+    type Err = ParseComplexError;
+
+    /// Parses `a +/- bi`; `ai +/- b`; `a`; or `bi` where `a` and `b` are of type `T`
+    fn from_str(s: &str) -> Result<Complex<T>, ParseComplexError>
+    {
+        let imag = match s.rfind('j') {
+            None => 'i',
+            _ => 'j'
+        };
+
+        let mut a = String::with_capacity(s.len());
+        let mut b = String::with_capacity(s.len());
+
+        let mut first = true;
+        let mut char_indices = s.char_indices();
+        let mut pc = ' ';
+
+        loop {
+            match char_indices.next() {
+                Some(t) => {
+                    let i = t.0;
+                    let cc = t.1;
+                    if cc == '+' {
+                        // handle exponents
+                        if pc != 'e' && pc != 'E' {
+                            first = false;
+                            // don't carry '+' over into b
+                            pc = ' ';
+                            continue;
+                        }
+                    } else if cc == '-' {
+                        // handle exponents and negative numbers
+                        if i > 0 && pc != 'e' && pc != 'E' {
+                            first = false;
+                            // DO carry '-' over into b
+                        }
+                    }
+
+                    if !first && cc == ' ' && pc == '-' {
+                        // ignore whitespace between minus sign and next number
+                        continue;
+                    }
+
+                    if first {
+                        a.push(cc);
+                    } else {
+                        b.push(cc);
+                    }
+
+                    pc = cc;
+                },
+                None => break,
+            }
+        }
+
+        a = a.trim_right().to_string();
+        b = b.trim_left().to_string();
+
+        if b.is_empty() {
+            b = match a.rfind(imag) {
+                None => "0i".to_string(),
+                _ => "0".to_string()
+            };
+        }
+
+        let re = match a.rfind(imag) {
+            None => {
+                try!(T::from_str(&a)
+                    .map_err(|_| ParseComplexError { kind: ComplexErrorKind::ParseError }))
+            },
+            _ => {
+                try!(T::from_str(&b)
+                    .map_err(|_| ParseComplexError { kind: ComplexErrorKind::ParseError }))
+            }
+        };
```

Comment:
I think rather than duplicating these parsing blocks, it would be better to just normalize `a` as real component and `b` as the imaginary beforehand.  Just swap them if needed, then make sure they're not both real or both imaginary, and then parse each piece.